File: templates/0-latest/js/solution.test.ts
import { Solution } from './solution';

const solution = new Solution();
const input = solution.readInput();
interface TestCase {
    input: string;
    args1?: any[];
    args2?: any[];
    want1: any;
    want2: any;
}

describe(`Day ${solution.day}`, () => {
    const tests: TestCase[] = [
        {
            input: `1000`,
            want1: 'TBD',
            want2: 'TBD',
        },
        {
            input,
            want1: 'TBD',
            want2: 'TBD',
        },
    ];

    test.each(tests)('Part 1 - Test %#', ({ input, want1, args1 }) => {
        expect(solution.part1(input, args1)).toEqual(want1);
    });

    test.each(tests)('Part 2 - Test %#', ({ input, want2, args2 }) => {
        expect(solution.part2(input, args2)).toEqual(want2);
    });
});

File: src/solutions/2023/11/go/solution/solution_test.go
package solution

import (
	"fmt"
	"reflect"
	"testing"
)

var (
	s     = New()
	input = s.ReadInputFromTests()
)

var tests = []struct {
	input string
	args1 []interface{}
	args2 []interface{}
	want1 interface{}
	want2 interface{}
}{
	{
		input: `...#......
.......#..
#.........
..........
......#...
.#........
.........#
..........
.......#..
#...#.....`,
		want1: 374,
		want2: 82000210,
	},
	{
		input: input,
		want1: 10885634,
		want2: 707505470642,
	},
}

func TestPart1(t *testing.T) {
	for i, tt := range tests {
		t.Run(fmt.Sprintf("Test_%d", i), func(t *testing.T) {
			if got := s.Part1(tt.input, tt.args1...); !reflect.DeepEqual(got, tt.want1) {
				t.Errorf("Part1() = %v, want %v", got, tt.want1)
			}
		})
	}
}

func TestPart2(t *testing.T) {
	for i, tt := range tests {
		t.Run(fmt.Sprintf("Test_%d", i), func(t *testing.T) {
			if got := s.Part2(tt.input, tt.args2...); !reflect.DeepEqual(got, tt.want2) {
				t.Errorf("Part2() = %v, want %v", got, tt.want2)
			}
		})
	}
}

func Test_getTotalDistance(t *testing.T) {
	input := `...#......
.......#..
#.........
..........
......#...
.#........
.........#
..........
.......#..
#...#.....`

	var tests = []struct {
		size int
		want int
	}{
		{size: 2, want: 374},
		{size: 10, want: 1030},
		{size: 100, want: 8410},
	}

	for i, tt := range tests {
		t.Run(fmt.Sprintf("Test_%d", i), func(t *testing.T) {
			if got := getTotalDistance(input, tt.size); !reflect.DeepEqual(got, tt.want) {
				t.Errorf("getTotalDistance() = %v, want %v", got, tt.want)
			}
		})
	}
}

func BenchmarkSolution_Part1(b *testing.B) {
	for i := 0; i < b.N; i++ {
		s.Part1(input)
	}
}

func BenchmarkSolution_Part2(b *testing.B) {
	for i := 0; i < b.N; i++ {
		s.Part2(input)
	}
}
